presto-hive/src/main/java/com/facebook/presto/hive/util/AsyncQueue.java

```diff
@@ -15,101 +15,144 @@
+    private final int softMaxQueueSize;
+
+    @GuardedBy("this")
+    private final Queue<T> elements;
+    // This future is completed when the queue transitions from empty to not. But it will be replaced by a new instance of future immediately.
+    @GuardedBy("this")
+    private CompletableFuture<?> notEmptySignal = new CompletableFuture<>();
+    // This future is completed when the queue transitions from full to not. But it will be replaced by a new instance of future immediately.
+    @GuardedBy("this")
+    private CompletableFuture<?> notFullSignal = new CompletableFuture<>();
+    @GuardedBy("this")
+    private boolean finished = false;
+
+    public AsyncQueue(int softMaxQueueSize)
     {
-        requireNonNull(element, "element is null");
-        if (finished.get()) {
-            return;
-        }
-        queue.add(element);
-        notEmptyFuture.get().complete(null);
+        checkArgument(softMaxQueueSize >= 1, "softMaxQueueSize must be at least 1");
+        this.softMaxQueueSize = softMaxQueueSize;
+        this.elements = new ArrayDeque<>(softMaxQueueSize * 2);
     }
 
-    /**
-     * Gets a batch of elements from the queue.  If queue is not empty, a
-     * completed future containing the queued elements is returned.  If the
-     * queue is empty, an incomplete future is returned that completes when
-     * an element is added.
-     * <p>
-     * It is possible that an empty list will be returned as another reader
-     * may have consumed all of the data.  A caller should read until, the
-     * {@code isFinished} method returns {@code true}.
-     *
-     * @param maxSize the maximum number of elements to return.
-     */
-    public CompletableFuture<List<T>> getBatchAsync(int maxSize)
+    public synchronized int size()
     {
-        return notEmptyFuture.get().thenApply(x -> getBatch(maxSize));
+        return elements.size();
     }
 
-    private List<T> getBatch(int maxSize)
+    public synchronized boolean isFinished()
     {
-        // take up to maxSize elements from the queue
-        List<Object> elements = new ArrayList<>(maxSize);
-        queue.drainTo(elements, maxSize);
-
-        // if the queue is empty and the current future is finished, create a new one so
-        // a new readers can be notified when the queue has elements to read
-        if (queue.isEmpty() && !finished.get()) {
-            CompletableFuture<?> future = notEmptyFuture.get();
-            if (future.isDone()) {
-                notEmptyFuture.compareAndSet(future, new CompletableFuture<>());
+        return finished && elements.size() == 0;
+    }
+
+    public void finish()
+    {
+        CompletableFuture<?> futureToSignal = null;
+        synchronized (this) {
+            if (finished) {
+                return;
+            }
+            finished = true;
+
+            if (elements.size() == 0) {
+                futureToSignal = notFullSignal;
+                notFullSignal = new CompletableFuture<>();
```

Comment:
I think it would be safer to assign this to `COMPLETED_FUTURE_OF_NULL`, that way it's impossible for anyone to block on this queue